### ISSUE TITLE
Fibre-local storage

### ISSUE THREAD

**Issue by SquidDev** [ocaml-multicore/eio#253]
Many concurrency libraries - both in OCaml[^1] and other languages - have some form of fibre-local storage, allowing you to pass implicit context around async code. It would be nice to have similar functionality in eio.

## Motivation
I'm currently working on an eio-based project where many tasks are run in parallel. Each task produces a lot of log output, and so it would be useful to track which log messages are associated with each task.

One way to achieve this would be to pass an explicit context around, and use that when logging (such as with `logs`'s `~tag` parameter). However, this obviously doesn't allow you to track log messages in external libraries, limiting its usefulness.

One common solution (at least in other languages) is to have some global (thread/fibre-local) variable which stores the current task. This is then be used when processing logs to associate them with the current task.

## Other notes
I realise this API may not be desirable for eio, as much of the API is designed around avoiding implicit state. Thought it was worth opening an issue and getting some thoughts before spending too much time thinking about how a potential API would behave.

Would be happy to implement this if it would be useful/helpful.

[^1]: lwt's `Lwt.new_key` and async's `Async.Execution_context`. While lwt's API is deprecated, its (sometimes) confusing behaviour could hopefully be avoided in eio due to the separation of promises and fibres.

**Comment by talex5** [comment]
Yes, I agree. It probably just needs an hmap in the `fiber_context`, alongside the thread ID:
https://github.com/ocaml-multicore/eio/blob/48c14c9902a022518a08c5bd5c8160d711bd497c/lib_eio/core/cancel.ml#L25-L30

The main concern I have is about modular reasoning - ideally two components shouldn't be able to interact without first being introduced. e.g. in

```ocaml
let x = ref 0 in
let y = Foo.make x in
let z = Bar.make () in
```
we should be able to assume that `z` does not have access to `x`. However, if `Foo.make` can put `x` in a fiber-local variable then this doesn't hold.

I see a couple of reasonable approaches:
1. Put this API in the `Private` module, so people know the usual rules don't apply, or
2. Only allow setting fiber-local variables at fiber-creation time, which I think might solve it (needs more thought).

**Comment by SquidDev** [comment]
> Only allow setting fiber-local variables at fiber-creation time 

I think either that or a `with_context` function, which replaces the internal hmap for the duration of the function (much like `Cancel.sub`) would be good. I guess the two are largely equivalent and can be used to implement the other!

**Comment by SquidDev** [comment]
> Using [hmap](https://github.com/dbuenzli/hmap) would probably be fine.

I guess one thing worth mentioning here is that creating keys won't be domain/thread safe (as it just uses a non-atomic reference cell). I assume that's fine, as long as it's documented?

**Comment by dbuenzli** [comment]
Using an `Atomic.t` value for the uid counter doesn't seem egregious.

**Comment by SquidDev** [comment]
Added in #256
